### PR TITLE
upgrade to webpack 5

### PR DESCRIPTION
**Pull request by ntarocco**
This PR updates webpack to version 5. Further all other packages except react and i18n packages are updated to the newest version to. This PR uses the caret as the default versioning tool to hopefully make the development easier.

NOTES:

- build is working
- assets watch is working
- assets watch-module is NOT working

### CODE REVIEW

**Comment by utnapischtim** [comment]
NOTE: this PR should be squashed. It is not necessary to keep all separate commits.

**Comment by max-moser** [comment]
### Dependencies updated
Updated a few dependencies because:

* `react-invenio-forms` 2.4.0 (the latest version) [requires query-string 7](https://github.com/inveniosoftware/react-invenio-forms/blob/master/package.json#L66), which clashes with the query-string 8 requirement 
* `friendly-webpack-plugin` 1.7.0 (the latest version, [from 5 years ago](https://www.npmjs.com/package/friendly-errors-webpack-plugin)) requires webpack 2, 3 or 4
* `optimize-css-assets-webpack-plugin` 6.0.1 (the latest version) has a [note in its README to use css-minimizer-webpack-plugin for webpack 5](https://www.npmjs.com/package/optimize-css-assets-webpack-plugin) instead


### Webpack 5 build without issues

Together with the change outlined in [my comment over here](https://github.com/inveniosoftware/invenio-app-rdm/pull/2187#issuecomment-1546878198), this enables me to build the assets for a fresh `my-site` with Webpack 5 and no issues (`invenio-cli install`, `invenio-cli assets build`).

Versions:
* Node 16.20.0
* NPM: 8.19.4
* Invenio-App-RDM: 12.0.0b2.dev2


### Watch

`invenio-cli assets watch` works without issues (tested with an editable install of `invenio-previewer`).

### Watch-Module

Tested by changing the label of `AccordionField` (adding some suffix [after `label`](https://github.com/inveniosoftware/react-invenio-forms/blob/master/src/lib/forms/AccordionField.js#L39)).

The directory layout of `react-invenio-forms` when installed from NPM:
````
mmoser@Charlie ~/.local/share/virtualenvs/my-site-3xGYYmMm/var/instance/assets $ ll node_modules/react-invenio-forms/ -d
Permissions Size User    Group   Date Modified Name
drwxr-xr-x     - mmoser mmoser 14 May 16:27  node_modules/react-invenio-forms

mmoser@Charlie ~/.local/share/virtualenvs/my-site-3xGYYmMm/var/instance/assets $ ll node_modules/react-invenio-forms/
Permissions  Size User    Group   Date Modified Name
drwxr-xr-x      - mmoser mmoser 14 May 16:27  dist
.rw-r--r--  1.1Ki mmoser mmoser 14 May 16:27  LICENSE
.rw-r--r--  3.6Ki mmoser mmoser 14 May 16:27  package.json
.rw-r--r--    675 mmoser mmoser 14 May 16:27  README.md
````

The layout when installed via `invenio-cli assets watch-module --link /path/to/react-module`:
```
mmoser@Charlie ~/.local/share/virtualenvs/my-site-3xGYYmMm/var/instance/assets $ ll node_modules/react-invenio-forms -d
Permissions Size User    Group   Date Modified Name
lrwxrwxrwx    60 mmoser mmoser 14 May 16:36  node_modules/react-invenio-forms -> ../../../../../../../../rdm/modules/react-invenio-forms/dist

mmoser@Charlie ~/.local/share/virtualenvs/my-site-3xGYYmMm/var/instance/assets $ ll node_modules/react-invenio-forms
Permissions  Size User    Group   Date Modified Name
drwxr-xr-x      - mmoser mmoser 14 May 16:36  cjs
drwxr-xr-x      - mmoser mmoser 14 May 16:36  esm
.rw-r--r--  3.6Ki mmoser mmoser 14 May 16:36  package.json
````
In the latter case, changes to the source code of `react-invenio-forms` *will* trigger a rebuild for both `watch-module` and `watch`. However, the changes don't seem to be compiled properly into the finished product, as the labels never changed on the deposit form in a running dev `my-site`.

File: invenio_assets/assets/build/webpack.config.js [hunk 1]
```diff
@@ -202,16 +221,22 @@ var webpackConfig = {
...
+      filename: path.join(config.build.assetsPath, "manifest.json"),
       publicPath: config.build.assetsURL,
     }),
   ],
   performance: { hints: false },
+  snapshot: {
+    managedPaths: [],
+  },
+  watchOptions: {
+    followSymlinks: true,
```
Comment (ntarocco) [inline invenio_assets/assets/build/webpack.config.js:233]: Can we add comments around, even link to GitHub issues, with explanations? Some of the options are obscure! :)

File: invenio_assets/assets/package.json [hunk 2]
```diff
@@ -7,62 +7,69 @@
   "scripts": {
     "start": "NODE_PRESERVE_SYMLINKS=1 NODE_ENV=development webpack --watch --progress --config ./build/webpack.config.js",
     "build": "NODE_PRESERVE_SYMLINKS=1 NODE_ENV=production webpack --config ./build/webpack.config.js",
-    "postinstalls": "patch-package"
+    "postinstall": "patch-package"
+  },
+  "dependencies": {
```
Comment (ntarocco) [inline invenio_assets/assets/package.json:12]: move this to `devDependencies`?

**Review by ntarocco — COMMENTED** [review]
Thanks a lot for this work!

**Comment by max-moser** [comment]
### Update on the `watch-module` :heavy_check_mark: 

With the configuration suggestion from [the last comment](https://github.com/webpack/webpack/issues/11612#issuecomment-1448208868) on [webpack issue 11612](https://github.com/webpack/webpack/issues/11612), the `watch-module` command works as expected again.

A temporary solution was to disable the cache (`cache: false` in the webpack config), but this caused very slow rebuilds on every change, but this isn't necessary anymore.

File: invenio_assets/assets/build/webpack.config.js (hunk 1 above)
Comment (max-moser) [reply invenio_assets/assets/build/webpack.config.js:235]: it's on my to-do list, after a brief break on frontend :D

**Pull request merged**